File: java/leetcode_grind/Day447.java
package leetcode_grind;

import java.util.ArrayList;
import java.util.List;

public class Day447 {
    // https://leetcode.com/problems/sequential-digits/description/
    static class Solution1 {
        public List<Integer> sequentialDigits(int low, int high) {
            String sample = "123456789";
            int n = 10;

            List<Integer> ans = new ArrayList<>();
            int lowLen = String.valueOf(low).length();
            int highLen = String.valueOf(high).length();

            for (int length = lowLen; length < highLen + 1; ++length) {
                for (int start = 0; start < n - length; ++start) {
                    int end = start + length;
                    int num = Integer.parseInt(sample.substring(start, end));

                    if (low <= num && num <= high) {
                        ans.add(num);
                    }
                }
            }

            return ans;
        }
    }
}

File: src/leetcode_grind/day_447.rs
// https://leetcode.com/problems/sequential-digits/description/
pub fn sequential_digits(low: i32, high: i32) -> Vec<i32> {
    fn build_el(dn: i32) -> i32 {
        let mut ans = 0;
        for i in 1..=dn {
            ans = ans * 10 + i;
        }
        ans
    }

    fn build_ones(dn: i32) -> i32 {
        let mut ans = 0;
        for _ in 0..dn {
            ans = ans * 10 + 1;
        }
        ans
    }

    let mut len = 2;
    let mut ones = build_ones(len);
    let mut el = build_el(2);
    let mut ans = vec![];

    while el <= high {
        if el >= low {
            ans.push(el);
        }
        if el % 10 == 9 {
            len += 1;
            el = build_el(len);
            ones = build_ones(len);
        } else {
            el += ones;
        }
    }

    ans
}
